### ISSUE TITLE
Diagram of modules

### ISSUE THREAD

**Issue by Kubuxu** [ipfs/go-ipfs#2452]
go-ipfs is big project. My biggest struggle while getting into it was finding my way through it.

A diagram of modules, how they communicate, in which directories they are contained and they functionalities, possibly with data flow paths would be great "getting into" helper.

cjdns has very basic version of that but it lacks a lot (also its codebase is much smaller and more consistent, it is C after all) other info that would be useful.

![](https://raw.githubusercontent.com/cjdelisle/cjdns/master/doc/CjdnsModules.png)

It also should be easy to update so it doesn't stall as it did in case of cjdns. There were some changes made that were never included into this diagram.

**Comment by whyrusleeping** [comment]
@Kubuxu do you know of any software that could be used to generate these graphs nicely? I can definitely lay down the groundwork of building things out

**Comment by Kubuxu** [comment]
Unfortunately not, nothing I ever used/tired to use was good.

**Comment by whyrusleeping** [comment]
hrm... microsoft visio is nice, but very much not free. Might be worth us trying something from this list: https://www.maketecheasier.com/5-best-free-alternatives-to-microsoft-visio/

**Comment by dignifiedquire** [comment]
http://melp.nl/2013/08/flow-charts-in-code-enter-graphviz-and-the-dot-language/

**Comment by whyrusleeping** [comment]
hrm... i do really love dot... it just doesnt have a nice GUI

**Comment by dignifiedquire** [comment]
GUI??? who needs a gui when you can write it in nice code, check that code into git and generate awesome graphs from it :P

**Comment by MichaelMure** [comment]
Maybe PlantUML ? http://plantuml.com/state.html

**Comment by lgierth** [comment]
I think we're looking for something like the diagrams in https://github.com/ipfs/js-ipfs, and maybe highlighting the used components for a few use cases -- the diagram tooling is secondary.

**Comment by dignifiedquire** [comment]
I agree on the tooling being secondary, if you post photos of handdrawn stuff or ascii art that's totally fine, and someone else can then turn them into a nicer format

**Comment by klivin** [comment]
I highly recommend draw.io, i recently jumped from omnigraffle, and it's fastest, best diagramming tool I've seen.  I saw this diagram of Ethereum which lead me here:
https://ethereum.stackexchange.com/questions/268/ethereum-block-architecture

I'd love love love to see a diagram of IPFS that's easier to grok than the Ethereum Proof of Work mechanism drawing.
